[ucd] Extend cross-component tests: GeneralCategory vs. BidiClass

calum opened the pull request:
This PR is my first attempt at closing #43. I have added 10 tests to the `category_tests.rs` test file.

I also corrected a typo in `category.rs` which was causing one of my tests to fail. 

Any feedback would be appreciated and I am happy to make changes.

UPDATE:
I removed my test for Bidi types OtherNeutral since from the [table of Bidi character types](http://www.unicode.org/reports/tr9/#Table_Bidirectional_Character_Types) this includes the General_Category types:

> All other characters, including OBJECT REPLACEMENT CHARACTER

Which is too general for a test - do you agree?

behnam commented:
Thanks, @calum, for the PR! I'll do a review in the next 24hrs.

In the meanwhile, would you please make sure your `master` branch is up-to-date and your diff is rebased on top of the latest master. After rebase, you can use `git push -f` to enforce the non-fast-forward commit to github.

Also, for commit and issue/PR titles, we follow this style:
```
[<area>] Fix something
[<area>] Add another thing
```
where, the `<area>` is either a UNIC component (like `ucd` or `ucd/category`/`ucd-category`) or other things like `tools`, `travis`, etc. This is not yet mentioned in `CONTRIBUTING.md`, but we should fix soon.

CAD97 commented on unic/ucd/tests/category_tests.rs:
```diff
@@ -18,8 +18,112 @@ extern crate unic_utils;
+/// `normal::is_combining_mark` and `GeneralCategory::is_mark()` are expected to return
+/// the same results.
+#[test]
+fn test_gen_cat_against_normal() {
```
After #57, this may be not testing the actual `ucd-normal` table for `is_combining_mark` but be calling on the implementation that calls back into `ucd-category`, because [I set `unic-ucd` to turn on the normal feature `unic-ucd-category`](https://github.com/behnam/rust-unic/pull/57/files#diff-588a4bcdb9c391c05ffe2270763cef14R22).

This test will need to either work around that or live instead in unic-normal. @behnam?

CAD97 commented on unic/ucd/tests/category_tests.rs:
```diff
@@ -18,8 +18,112 @@ extern crate unic_utils;
+            assert!(
+                GC::of(cp).is_symbol()          ||
+                GC::of(cp) == GC::Unassigned    ||
+                GC::of(cp) == GC::OtherPunctuation
```
Your test and comments don't agree here.

The doc comment says `bc=ET ⊆ ( gc=S ∪ gc=Cn ∪ gc=Po )`,  
The implementation comment says `bc=ET ⊆ ( gc=S ∪ gc=No ∪ gc=Po )`,  
The implementation says `bc=Et ⊆ ( gc=S ∪ gc=Cn ∪ gc=Po )`.

Which is it?

For what it's worth, I don't _think_ `gc=Cn` have any properties. I could be wrong though, I haven't studies the bidi data files.

calum commented:
My rebase keeps going wrong. So when I run ```$ git rebase upstream/master``` I get this output:

```sh
First, rewinding head to replay your work on top of it...
Applying: working on it
Applying: Added test to check that normal::is_combining_mark and GC::is_mark agree
Applying: [utils] Merge unic-utils-codepoints src into unic-utils
Using index info to reconstruct a base tree...
M	unic/ucd/age/Cargo.toml
M	unic/ucd/age/tests/age_tests.rs
M	unic/ucd/normal/Cargo.toml
M	unic/ucd/normal/tests/conformance_tests.rs
M	unic/utils/Cargo.toml
A	unic/utils/codepoints/Cargo.toml
A	unic/utils/codepoints/src/codepoints.rs
A	unic/utils/codepoints/src/lib.rs
M	unic/utils/src/lib.rs
Falling back to patching base and 3-way merge...
CONFLICT (rename/delete): unic/ucd/core/Cargo.toml deleted in [utils] Merge unic-utils-codepoints src into unic-utils and renamed in 4d5712f1e797eeb71489b140fb84d361be4d2502. Version 4d5712f1e797eeb71489b140fb84d361be4d2502 of unic/ucd/core/Cargo.toml left in tree.
CONFLICT (rename/delete): unic/idna/mapping/src/lib.rs deleted in [utils] Merge unic-utils-codepoints src into unic-utils and renamed in 4d5712f1e797eeb71489b140fb84d361be4d2502. Version 4d5712f1e797eeb71489b140fb84d361be4d2502 of unic/idna/mapping/src/lib.rs left in tree.
error: Failed to merge in the changes.
Patch failed at 0003 [utils] Merge unic-utils-codepoints src into unic-utils
The copy of the patch that failed is found in: .git/rebase-apply/patch

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
```

Then when I run ```$ git status```:
```sh
rebase in progress; onto f6349c7
You are currently rebasing branch '43_extending_tests' on 'f6349c7'.
  (fix conflicts and then run "git rebase --continue")
  (use "git rebase --skip" to skip this patch)
  (use "git rebase --abort" to check out the original branch)

Unmerged paths:
  (use "git reset HEAD <file>..." to unstage)
  (use "git add <file>..." to mark resolution)

	added by us:     ../../idna/mapping/src/lib.rs
	added by us:     ../core/Cargo.toml

no changes added to commit (use "git add" and/or "git commit -a")

```

But I'm confused because I never changed those files in any of my commits? What's the best way to get around this?

@CAD97 Thanks for pointing out those typos. I've corrected them now. While writing the tests I found that quite a lot of the EuropeanTerminator were coming out as GC::Unassigned. I used the [table](http://www.unicode.org/reports/tr9/#Table_Bidirectional_Character_Types) to get a rough idea of what was expected, then wrote some ```println!()``` lines to see what was coming out,  then wrote the tests. So if a test doesn't look right to you, then I could remove it. I read that the Bidi types don't map exactly over to the Category types.

@behnam Thanks for the help with the naming. I have fixed the name of this pull request.

behnam commented on unic/ucd/tests/category_tests.rs:
```diff
@@ -18,8 +18,112 @@ extern crate unic_utils;
+            assert!(
+                GC::of(cp) == GC::Format            ||
+                GC::of(cp) == GC::OtherNumber       ||
+                GC::of(cp) == GC::OtherPunctuation  ||
```
Instead of `||` and `==`, you can also use `matches!()` which has a shorter syntax, and the `of()` function would get called only once.

behnam commented on unic/ucd/tests/category_tests.rs:
```diff
@@ -18,8 +18,112 @@ extern crate unic_utils;
+    }
+}
+
+/// `Bidi_Class=AN := General_Category in { Cf (Format), No (OtherNumber),
```
In general, I feel like the comments are not that helpful here and make it a bit more complicated. I think the implementation can be enough, and if there's any comment, it can be an exact copy-paste of the source, so when reading the text, we can see how it's implemented. Right now, there are two levels of interpretation, which is making me confused a bit.

In this example, using `matches` macro would give you the exact same reading in the code, as you have in the comment.

behnam commented on unic/ucd/tests/category_tests.rs:
```diff
@@ -18,8 +18,112 @@ extern crate unic_utils;
+/// <http://www.unicode.org/reports/tr9/#AN>
+#[test]
+fn test_bidi_an_against_gen_cat() {
+    // Every BC::ArabicNumber must be a GC::Format, GC::OtherNumber, GC::OtherPunctuation,
```
This comment is also not adding anything, and just a copy of the logic in the code.

behnam commented on unic/ucd/tests/category_tests.rs:
```diff
@@ -18,8 +18,112 @@ extern crate unic_utils;
+/// `Bidi_Class=AN := General_Category in { Cf (Format), No (OtherNumber),
+///                                         Po (OtherPunctuation), Nd (DecimalNumber) }`
+///
+/// <http://www.unicode.org/reports/tr9/#AN>
```
This reference URLs are useful, though.

behnam commented on unic/ucd/tests/category_tests.rs:
```diff
@@ -18,8 +18,112 @@ extern crate unic_utils;
 
 use unic_ucd::bidi::BidiClass;
 use unic_ucd::normal::is_combining_mark;
+use unic_ucd::category::GeneralCategory as GC;
```
Feel free to also `use ...::BidiClass as BC`, if you think makes the lines shorter and removes the clutter.

behnam commented on unic/ucd/tests/category_tests.rs:
```diff
@@ -40,3 +144,67 @@ fn test_bidi_nsm_against_gen_cat() {
+
+        // Every !GC::is_other() must not be an BoundaryNeutral
+        if !GC::of(cp).is_other() {
+            assert_ne!(BidiClass::of(cp), BidiClass::BoundaryNeutral);
```
I think this assertion fails if and only if the one above fails. If so, we can drop it.

calum commented:
@behnam  Thanks for your feedback. I removed the unnecessary comments and some lines of code. I wanted to use the `matches!` macro, but I wasn't able to use it. Do you want me to use an external crate? I couldn't find the macro on the rust docs: (https://doc.rust-lang.org/std/#macros).

behnam commented:
We use the in the code link this: https://github.com/behnam/rust-unic/blob/master/unic/bidi/src/implicit.rs#L238
```
fn is_NI(class: BidiClass) -> bool {
    matches!(class, B | S | WS | ON | FSI | LRI | RLI | PDI)
}
```

First, you need to add it to the manifest (`Cargo.toml`) like this: <https://github.com/behnam/rust-unic/blob/master/unic/bidi/Cargo.toml#L26>. In this case, since it's only used in the tests, it goes under `dev-dependencies`, like this:
```
[dev-dependencies]
matches = "0.1"
```

Also, on top of the crate (the test file, in this case), you need to tell rustc to use the external resource, like this: <https://github.com/behnam/rust-unic/blob/master/unic/bidi/src/lib.rs#L72>.  Note that you need to explicitly enable macro usage:
```
#[macro_use]
extern crate matches;
```

(PS. The reason for explicit macro usage is lack of namespace system for macros. There's some discussion online about this already.)

Let me know if that resolves the issue. If not, I'll be glad to help if you provide more info.

behnam commented on unic/ucd/tests/category_tests.rs:
```diff
@@ -16,10 +16,85 @@ extern crate unic_ucd;
+/// `normal::is_combining_mark` and `GeneralCategory::is_mark()` are expected to return
+/// the same results.
+#[test]
+fn test_gen_cat_against_normal() {
```
I have added this test to the more-specific component (`ucd-normal`) in <https://github.com/behnam/rust-unic/pull/65>, so you can now drop it from here and have this file only focus on tests for bidi-class against general-category.

calum commented:
@behnam I have made further changes: dropping the ucd-normal test; changing the assertions to use the `matches!` macro.

I'm happy to make any further changes if you require them.
